instrumentation/jmx-metrics/javaagent/README.md

```diff
@@ -221,6 +221,54 @@ Thus, the above definitions will create several metrics, named `my.kafka.streams
 
 The metric descriptions will remain undefined, unless they are provided by the queried MBeans.
 
+### State Metrics
+
+Some JMX attributes expose current state as a non-numeric MBean attribute, in order to capture those as metrics it is recommended to use the special `state` metric type.
+For example, with Tomcat connector, the `Catalina:type=Connector,port=*` MBean has `stateName` (of type `String`), we can define the following rule:
+
+```yaml
+---
+rules:
+  - bean: Catalina:type=Connector,port=*
+    mapping:
+      stateName:
+        type: state
+        metric: tomcat.connector.state
+        metricAttribute:
+          port: param(port)
+          state:
+            ok: STARTED
+            failed: [STOPPED,FAILED]
+            degraded: '*'
+```
+
+For a given value of `port`, let's say `8080` This will capture the `tomcat.connector.state` metric of type `updowncounter` with value `0` or `1` and the `state` metric attribute will have a value in [`ok`,`failed`,`degraded`].
+For every sample, 3 metrics will be captured for each value of `state` depending on the value of `stateName`:
+
+When `stateName` = `STARTED`, we have:
+
+- `tomcat.connector.state` value = `1`, attributes `port` = `8080` and `state` = `ok`
+- `tomcat.connector.state` value = `0`, attributes `port` = `8080` and `state` = `failed`
+- `tomcat.connector.state` value = `0`, attributes `port` = `8080` and `state` = `degraded`
+
+When `stateName` = `STOPPED` or `FAILED`, we have:
+
+- `tomcat.connector.state` value = `0`, attributes `port` = `8080` and `state` = `ok`
+- `tomcat.connector.state` value = `1`, attributes `port` = `8080` and `state` = `failed`
+- `tomcat.connector.state` value = `0`, attributes `port` = `8080` and `state` = `degraded`
+
+For other values of `stateName`, we have:
+
+- `tomcat.connector.state` value = `0`, attributes `port` = `8080` and `state` = `ok`
+- `tomcat.connector.state` value = `0`, attributes `port` = `8080` and `state` = `failed`
+- `tomcat.connector.state` value = `1`, attributes `port` = `8080` and `state` = `degraded`
+
+Each state key can be mapped to one or more values of the MBean attribute using:
+- a string literal or a string array
+- a `*` wildcard to provide default option and avoid enumerating all values.
+
+Exactly one wildcard is expected to be present in the mapping to ensure all possible values of the MBean attribute can be mapped to a state key.
```

Comment:
My understanding is that the wildcard is not mandatory, just a recommended practice.

Reply:
No, it needs to be mandatory to ensure every possible state is covered and mapped to a "state key", in a sense it's similar to having all possible values of an enum covered in a switch statement.

The value is commonly presented as a `String` in the JMX interface, but we don't have any guarantees on the cardinality, even if in practice it's often the string representation of an enum (like in the Tomcat example).

Reply:
I'm still not not convinced. Yes, there's some similarity to a switch statement, but it is weak argument. Some people might care for only some states out of many possible. Some MBeans may be known for producing just two distinct states (like "on" or "off"), and the `default` clauses would look very artificial.
Furthermore, the cardinality of the metric does not change when `default` is present. The only thing guaranteed by the mandatory `default` entries is that the metric value when aggregated over all metric attributes representing state will be `1` - a feature of no real value, IMHO.

Reply:
The default value can be one of already-mapped keys, it does not have to be a separate artificial state key, the idea is that we need to make it explicit to avoid any ambiguity in the mapping. I should probably add another example for this use-case as well.

For example, with an `RUNNING|STOPPED` value mapped to `on|off` you could use:

```
  on: RUNNING
  off: '*'
```

or when including the known value:

```
  on: RUNNING
  off: [STOPPED,'*']
```

An alternative option could be to "pick the last key" as default, but this behavior might provide surprising results and could lead to capturing invalid metrics.